drivers/backup/portworx/portworx.go

```diff
@@ -204,6 +206,44 @@ func (p *portworx) DeleteBackup(req *api.BackupDeleteRequest) (*api.BackupDelete
 	return p.backupManager.Delete(context.Background(), req)
 }
 
+// WaitForBackupCompletion waits for backup to complete successfully
+// or till timeout is reached. API should poll every `timeBeforeRetry` duration
+func (p *portworx) WaitForBackupCompletion(req *api.BackupInspectRequest,
+	timeout time.Duration, timeBeforeRetry time.Duration) error {
+	var backupError error
+	f := func() (interface{}, bool, error) {
+		inspectBkpResp, err := p.backupManager.Inspect(context.Background(), req)
+		if err != nil {
+			// Error occured, just retry
+			return nil, true, err
+		}
+
+		// Check if backup status is complete
+		currentStatus := inspectBkpResp.GetBackup().GetStatus().GetStatus()
+		if currentStatus == api.BackupInfo_StatusInfo_Success {
+			// If backup is complete, dont retry again
+			return nil, false, nil
+		} else if currentStatus == api.BackupInfo_StatusInfo_Failed ||
+			currentStatus == api.BackupInfo_StatusInfo_Aborted ||
+			currentStatus == api.BackupInfo_StatusInfo_Invalid {
+			backupError = fmt.Errorf("backup [%v] is in [%s] state",
+				req.GetName(), currentStatus)
+			return nil, false, backupError
+		}
+		return nil,
+			true,
+			fmt.Errorf("backup [%v] is in [%s] state. Waiting to become Complete",
+				req.GetName(), currentStatus)
+	}
+
+	_, err := task.DoRetryWithTimeout(f, timeout, timeBeforeRetry)
```

Comment:
What is the reason of using backupErr through closure? Is there a case when returned error is lost inside task  code?

Reply:
@stgleb yes, the returned error is getting lost inside the task code. In case of failed backups, `nil` error is getting returned.

Reply:
Wow,that is wierd. I'll update my code for that.